### ISSUE TITLE
revalidateTag doesn't work after page navigation

### ISSUE THREAD

**Issue by HellYeahOmg** [vercel/next.js#68028]
### Link to the code that reproduces this issue

https://github.com/HellYeahOmg/revalidate-test/tree/long-async-request

### To Reproduce

1. Start the app
2. On the main page submit the form. POST route handler has a 5 seconds timeout
3. Navigate to the blank page from the header, wait for the response to finish
4. Navigate back to the main page

### Current vs. Expected behavior

On navigate back to the main page i expect to see updated data.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.4.0: Fri Mar 15 00:10:42 PDT 2024; root:xnu-10063.101.17~1/RELEASE_ARM64_T6000
  Available memory (MB): 16384
  Available CPU cores: 10
Binaries:
  Node: 18.17.1
  npm: 9.6.7
  Yarn: N/A
  pnpm: 9.4.0
Relevant Packages:
  next: 14.2.5 // Latest available version is detected (14.2.5).
  eslint-config-next: N/A
  react: 18.3.1
  react-dom: 18.3.1
  typescript: N/A
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure, Navigation

### Which stage(s) are affected? (Select all that apply)

next dev (local)

### Additional context

First I encountered in production a server action error on 13.5.6 version of next with this code (i can provide a github repo if needed). Then i bumped the version to the current release, and while the error was fixed, caching behaviour doesn't work.

**Comment by AliRazaDev27** [comment]
No sure what you mean by Navigate to the blank page in the header, still if you mean you are waiting on the form page, then you can use revalidatePath instead of revalidateTag as it does not have any real logical effect on your code.
Also you have to assign a tag in  your fetch request to properly make it work. read the docs. i will try opening a pull reqest though.

**Comment by HellYeahOmg** [comment]
It's literally called BlankPage in the code. 
`<Link href="/blank">Blank page</Link>`

The requests also does have a tag:
`  const data = await fetch('http://localhost:3000/api', { cache: 'no-store', next: { tags: ['data'] } })
`

Moreover, neither revalidatePath does work.

**Comment by AliRazaDev27** [comment]
Check your Pull requests.

**Comment by HellYeahOmg** [comment]
@AliRazaDev27 thank you, but you misunderstand the question. This issue is about revalidateTag (revalidatePath) not updating the cache while navigating to another page. 

I made the repo specifically to test it, there's no need to circumvent the scenario, because it is what I experience on my production. On the prod users can upload huge files which take some time, and they are expected to navigate to other pages while it happens, and when they go back they should see the updated results.

**Comment by HellYeahOmg** [comment]
There’s no need to use a storage. Unless you make any changes to the code it’s there, which is enough to test the case. 

And yes, I’m not using any cache and yet the results are cached, that’s this issue is about :)

**Comment by AliRazaDev27** [comment]
Results are not being cached but reset.

**Comment by AliRazaDev27** [comment]
First issue is the your src/app/page.js itself is being cached. so GET request is not being called on page navigation, let me look around a bit.

**Comment by AliRazaDev27** [comment]
Here is the issue, you page is being cached in Router cache, which is client side.
https://nextjs.org/docs/app/building-your-application/caching#router-cache
The only way is to either do a router.refresh, revalidate or set a config option.
for revalidate, you have to either remove cache:"no store" or set cache:"force-cache" so that the revalidateTag can work properly,
just tested on my machine and it is working.
https://nextjs.org/docs/app/api-reference/next-config-js/staleTimes
const nextConfig = {
  experimental: {
    staleTimes: {
      dynamic: 30,
      static: 180,
    },
  },
}

**Comment by HellYeahOmg** [comment]
The docs clearly state that we can invalidate router cache by calling router.refresh, revalidatePath, or revalidateTag, yet it's not helping in that case.

**Comment by AliRazaDev27** [comment]
Move revalidateTag from client to server. On page navigation, the revalidateTag function  on client does not work.
Currently with respect to revalidateTag, you can use 2 different approaches. first is put your revalidateData on your blank page also, which works but is not the best approach if you a lot of different pages you want to switch between,
second is to use revalidateTag on the server (see pull request), you can still use router.refresh(), or revalidatePath, since the issue is with Router cache (client side cache)

**Comment by AliRazaDev27** [comment]
Check your code for thrown errors, in my case, just putting console.log() in revalidateData() server action and checking the behavior difference b/w on page revalidation and on navigation revalidation suggest the revalidateData call is not executed properly.Also it throws and error in my server console.
TypeError: __webpack_modules__[moduleId] is not a function.
It looks like the function reference is lost during page navigation.

**Comment by AliRazaDev27** [comment]
No, it is not a bug , since the issue is with the Router cache (client-side cache) and currently it can not be opted out.So you have to tell the browser to purge the Router cache somehow, there are currently 3 ways (+1 experimental which i mentioned before). router.refresh, revalidateTag and revalidatePath, The problem you are facing is that you are waiting in a client component with await and then calling revalidateData(tag) but on page navigation during this client await, the call to revalidateData is lost, u can check whether the function is called in the background by using console.log or any other debugging method you prefer, which i found is not called on navigation, thus the browser is never asked to purge the Router cache, wtih revalidateTag, you can either call it from the server, so on navigation the call is not interrupted or from the client component you navigated to (though it is not recommended).

**Comment by marcoshernanz** [comment]
Thanks for the report. I can reproduce this on `14.2.5`, but not on the latest supported Next.js 15 or 16 releases, even with the old 30-second dynamic stale time restored. The delayed mutation and revalidation complete after navigation, and returning to the page shows fresh data. Since this is fixed in supported releases, I'm closing this. If it still reproduces after upgrading, please open a new issue with an updated reproduction.
